Aave V2 Treasury: Treasury Management Increase aUSDC holdings on Collector

efecarranza opened the pull request:
# Changelog

Create proposal to acquire aUSDC tokens in order to fund service providers.
Proposal 1 of 2.
Add test for swap.
Add sample script with transaction.

rustboyar commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.sol:
```diff
@@ -0,0 +1,90 @@
+ * @title Acquire more aUSDC to pay service providers
+ * @author Llama
+ * - Snapshot: https://governance.aave.com/t/arfc-deploy-ethereum-collector-contract/12205
+ * - Discussion: https://snapshot.org/#/aave.eth/proposal/0xb4141f12f7ec8e037e6320912b5673fcc5909457d9f6201c018d5c15e5aa5083
```
snapshot and discussion linked are mixed up

rustboyar commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.sol:
```diff
@@ -0,0 +1,90 @@
+contract AaveV2_Eth_ServiceProviders_20231907 is IProposalGenericExecutor {
+  using SafeERC20 for IERC20;
+
+  error InvalidCaller();
```
seems that this error is not used anywhere

rustboyar commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.sol:
```diff
@@ -0,0 +1,90 @@
+
+  error InvalidCaller();
+
+  address public constant ALLOWED_CALLER = 0xA519a7cE7B24333055781133B13532AEabfAC81b;
```
this constant is not not used

rustboyar commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.t.sol:
```diff
@@ -0,0 +1,66 @@
+
+    assertEq(IERC20(AaveV2EthereumAssets.USDC_A_TOKEN).balanceOf(address(payload)), 0);
+
+    assertGt(
```
could you please make assertApprox here as well, not just greater

rustboyar commented on src/AaveV2_Eth_ServiceProviders_20231907/ServiceProviders.md:
```diff
@@ -0,0 +1,26 @@
+
+## Motivation
+
+## Specification
```
missing section

rustboyar commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.sol:
```diff
@@ -0,0 +1,90 @@
+    IERC20(AaveV2EthereumAssets.USDT_A_TOKEN).safeApprove(MILKMAN, AMOUNT_USDT);
+    IERC20(AaveV2EthereumAssets.DAI_A_TOKEN).safeApprove(MILKMAN, AMOUNT_DAI);
+
+    IMilkman(MILKMAN).requestSwapExactTokensForTokens(
```
what if swap won't get executed? do we need the the ability to cancelSwap?

efecarranza replied:
@rustboyar because Milkman uses `msg.sender`, a call from the Short Executor would be the only one able to cancel a swap. This swap is highly unlikely to get stuck. We have two options until AaveCurator gets deployed (that will simplify many things going forward).

1. Deploy a new contract from the payload to do the swap and also be able to cancel
2. Do a new proposal to cancel if that happens

I'd do option 1 if we want to have the quickest fail-safe mechanism without having to go through governance. It's a small price to pay.

Really excited to have AaveCurator out finally so we are not having to re-invent the wheel every time.

efecarranza replied:
Updated!

rustboyar replied:
Hey! Can't see the update or it was related to smth other than the swap cancel?

efecarranza commented on src/AaveV2_Eth_ServiceProviders_20231907/ServiceProviders.md:
```diff
@@ -0,0 +1,26 @@
+
+This publication intends to swap several assets via Cowswap to USDC. The USDC is to be deposited in Aave v2 for aUSDC and then used to remunerate Service Providers via the various streaming contracts.
+
+## Motivation
```
so weird! i added this! 🤯  i'll re add

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.sol:
```diff
@@ -0,0 +1,87 @@
+
+    AaveV2Ethereum.COLLECTOR.transfer(AaveV2EthereumAssets.USDC_UNDERLYING, address(this), balance);
+
+    IERC20(AaveV2EthereumAssets.USDC_UNDERLYING).safeApprove(address(AaveV2Ethereum.POOL), balance);
```
Hi, SafeApprove is deprecated as it faces the same issues as the standard approval, so there's no real reason to use it here. It's also removed from updated OpenZeppelin libs.

efecarranza replied:
@Zer0dot quick question, how do you approve USDT without `safeApprove`? Thank you.

Zer0dot replied:
As this isn't really a problem, going to resolve this!

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.sol:
```diff
@@ -0,0 +1,87 @@
+
+    IMilkman(MILKMAN).requestSwapExactTokensForTokens(
+      AMOUNT_USDT,
+      IERC20(AaveV2EthereumAssets.USDT_A_TOKEN),
```
Maybe I'm missing something but isn't liquidity quite thin for aTokens? Do CoWSwap solvers account for this and include supplying assets in the trade execution?

Zer0dot replied:
Wait, there's Curve, nevermind lol.

efecarranza replied:
hey @Zer0dot COW Swap solvers actually do the flow of depositing, etc. so they account for all that.

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/TestTrade.s.sol:
```diff
@@ -0,0 +1,50 @@
+  address public constant MILKMAN = 0x11C76AD590ABDFFCD980afEC9ad951B160F02797;
+  address public constant CHAINLINK_PRICE_CHECKER = 0x4D2c3773E69cB69963bFd376e538eC754409ACFa;
+
+  // forge script src/AaveV2_Eth_ServiceProviders_20231907/TestTrade.s.sol:TestTrade --rpc-url mainnet --broadcast --private-key ${PRIVATE_KEY} --slow -vvvv
```
Can we have this in NatSpec at the top of the contract with an explanation as to what the contract is? It makes sense to have, but since it tests "in prod" it can be a bit confusing as to how it works.

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.t.sol:
```diff
@@ -0,0 +1,71 @@
+
+import {AaveV2_Eth_ServiceProviders_20231907} from './AaveV2_Eth_ServiceProviders_20231907.sol';
+
+contract AaveV2_Eth_ServiceProviders_20231907_Test is Test {
```
Just for readability purposes, consider declaring the tokens you need at the top of the test, instead of accessing the library constant and casting it each time. Not a big deal though.

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.t.sol:
```diff
@@ -0,0 +1,71 @@
+    assertApproxEqAbs(
+      IERC20(AaveV2EthereumAssets.USDC_A_TOKEN).balanceOf(address(AaveV2Ethereum.COLLECTOR)),
+      balanceBeforeAUSDC + balanceBeforeUSDC,
+      1e18
```
Why 1e18 while others are 1? Also, isn't v2 aUSDC 6 decimals?

efecarranza replied:
So I think it's weird doing approxEqAbs, because it's not exactly 1-to-1 when depositing. 
The difference is 180 USDC total, which is not a lot considering it's 290,000 getting deposited.
So that's why I had done `assertGt` previously. I can go either way, definitely 1e18 was a mistake.

As to why others are 1? There's always some rounding error and I'm withdrawing which is different.

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/AaveV2_Eth_ServiceProviders_20231907.sol:
```diff
@@ -0,0 +1,87 @@
+
+    data = abi.encode(paths, reverses);
+
+    return abi.encode(100, data); // 100 = 1% slippage
```
Why such a high slippage for a stable => stable swap? We are swapping nearly 2 MM$ and 1% is nearly 20k, is this really needed?

efecarranza replied:
It doesn't necessarily have to be 1%, that's as far as it can go and it was the number recommended by COW Swap. For the 80BAL20WETH trade I had done 1.5% and ended up trading at 1% slippage. It's the maximum, not what will be executed at.
I found a bug in their implementation so I'm waiting for a patch also.

Zer0dot replied:
Damn, that's a high slippage, but either way just a comment.

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/ServiceProviders.md:
```diff
@@ -0,0 +1,38 @@
+---
+title: Acquire more aUSDC to pay service providers
```
Title is not capitalized properly-- small thing, but worth fixing.

efecarranza commented:
@Zer0dot some questions so we can discuss how to move forward. thank you!

efecarranza commented:
@rustboyar @Zer0dot 

I've addressed all concerns and updated accordingly everything. 
COWSwap had to update their price checker so I didn't push before.

Please let me know if anything else needs attention.

Zer0dot commented on src/AaveV2_Eth_ServiceProviders_20231907/TestTrade.s.sol:
```diff
@@ -48,6 +50,6 @@ contract TestTrade is Script {
     data = abi.encode(paths, reverses);
 
-    return abi.encode(2000, data); // 2000 = 20% slippage just cause it's a super small amount
+    return abi.encode(3500, data); // 2000 = 20% slippage just cause it's a super small amount
```
Seems this changed for testing? Comment should probably be updated, anyway why is the slippage so high for testing, gas concerns?

efecarranza replied:
100 USDC is not much unfortunately, so need to do a big spread for it to be picked up and get killed on the slippage 'cause if not it doesn't even cover gas costs.

efecarranza replied:
It's not so much that it's only say going from 100 USDC to 99.5 aUSDC or whatever, the slippage tolerance has to account for gas costs also so that's why it ends up being bigger.

Zer0dot replied:
Fair enough, got it!
